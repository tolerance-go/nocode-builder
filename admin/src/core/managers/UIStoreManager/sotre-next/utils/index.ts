import { ProjectStructureTreeDataNode } from '@/types';
import { PayloadAction } from '@reduxjs/toolkit';
import { ProjectTreeStates } from '../slices/projectTree';

export const findProjectStructureTreeNode = (
  state: ProjectTreeStates,
  action: PayloadAction<string>,
): ProjectStructureTreeDataNode | null => {
  const key = action.payload;
  const findNode = (
    nodes: ProjectStructureTreeDataNode[],
  ): ProjectStructureTreeDataNode | null => {
    for (const node of nodes) {
      if (node.key === key) {
        return node;
      }
      if (node.children) {
        const found = findNode(node.children);
        if (found) {
          return found;
        }
      }
    }
    return null;
  };

  return findNode(state.projectStructureTreeData);
};
