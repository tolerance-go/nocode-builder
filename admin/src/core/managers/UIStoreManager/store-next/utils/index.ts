import { ProjectStructureTreeDataNode } from '@/types';
import { ProjectTreeStates } from '../slices/projectTree';

export const findProjectStructureTreeNode = (
  state: ProjectTreeStates,
  key: string,
): ProjectStructureTreeDataNode | null => {
  const findNode = (
    nodes: ProjectStructureTreeDataNode[],
  ): ProjectStructureTreeDataNode | null => {
    for (const node of nodes) {
      if (node.key === key) {
        return node;
      }
      if (node.children) {
        const found = findNode(node.children);
        if (found) {
          return found;
        }
      }
    }
    return null;
  };

  return findNode(state.projectStructureTreeData);
};
