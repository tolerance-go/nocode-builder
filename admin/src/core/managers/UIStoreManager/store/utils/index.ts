import { ProjectStructureTreeDataNode } from '@/types';
import { ProjectTreeStates } from '../slices/projectTree';

export const 查询项目树中的节点 = (
  state: ProjectTreeStates,
  key: string,
): ProjectStructureTreeDataNode | null => {
  const findNode = (
    nodes: ProjectStructureTreeDataNode[],
  ): ProjectStructureTreeDataNode | null => {
    for (const node of nodes) {
      if (node.key === key) {
        return node;
      }
      if (node.children) {
        const found = findNode(node.children);
        if (found) {
          return found;
        }
      }
    }
    return null;
  };

  return findNode(state.项目节点树);
};
