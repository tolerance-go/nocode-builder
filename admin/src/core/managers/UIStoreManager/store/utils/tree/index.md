设计一个`tree-model`的操作模型时，应该提供一组全面的API来支持创建、修改、查询和遍历树形结构。以下是一些建议的API，这些API旨在覆盖处理树形结构时的常见需求：

1. **初始化和配置**
   - `TreeModel(config)`: 构造函数，用于创建一个新的`TreeModel`实例。`config`参数用于自定义树的行为，如设置子节点的属性名称或比较函数。
2. **节点创建和解析**
   - `parse(model)`: 将给定的模型对象解析为树形结构的根节点。
   - `createNode(data)`: 创建一个新的节点，其中`data`是节点要保存的数据。
3. **节点操作**
   - `addChild(parent, child)`: 向指定的父节点添加一个新的子节点。
   - `addChildAtIndex(parent, child, index)`: 在指定的索引位置向父节点添加一个新的子节点。
   - `removeChild(parent, child)`: 从父节点移除一个子节点。
   - `setParent(child, newParent)`: 更改节点的父节点。
   - `replaceChild(parent, oldChild, newChild)`: 替换父节点下的一个子节点。
4. **查询和遍历**
   - `traverse(strategy, callback)`: 遍历树，可以指定遍历策略（如深度优先或广度优先）和回调函数。
   - `find(condition)`: 查找满足特定条件的第一个节点。
   - `findAll(condition)`: 查找所有满足特定条件的节点。
   - `getPathTo(node)`: 获取从根节点到指定节点的路径。
   - `getRoot()`: 获取树的根节点。
5. **排序和比较**
   - `sortChildren(node, comparator)`: 根据比较函数对节点的子节点进行排序。
6. **信息获取**
   - `getParent(node)`: 获取节点的父节点。
   - `getChildren(node)`: 获取节点的子节点。
   - `getData(node)`: 获取节点保存的数据。
   - `getIndex(node)`: 获取节点在其父节点的子节点列表中的索引。
7. **树结构操作**
   - `clone(node)`: 克隆一个节点或整个树。
   - `merge(t1, t2)`: 合并两棵树。
   - `split(node)`: 将一个节点及其子树从原树中分离出来。
8. **事件和通知**
   - `addEventListener(type, listener)`: 添加事件监听器，以便在树结构发生变化时得到通知。
   - `removeEventListener(type, listener)`: 移除事件监听器。
     这些API为处理树形结构提供了一个全面的工具集，可以满足多种不同的需求。根据实际应用场景，可能还需要添加或调整某些API以提供更具体的功能。
