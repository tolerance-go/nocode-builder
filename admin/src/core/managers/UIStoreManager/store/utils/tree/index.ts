export * from './effects';
export * from './misc';
