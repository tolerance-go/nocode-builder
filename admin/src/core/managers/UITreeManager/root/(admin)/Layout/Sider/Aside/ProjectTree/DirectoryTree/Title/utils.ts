const 标题格式是否有效 = (input: string): boolean => {
  // 不允许 Postgresql 中的特殊字符
  const regex = /^[^!@#$%^&*()_+\-={}|[\]\\:";'<>?,./]*$/;
  return regex.test(input);
};
// 标题输入框内容是否不合法
export const 标题是否有错 = (value: string): boolean => {
  return value.trim() === '' || !标题格式是否有效(value);
};
