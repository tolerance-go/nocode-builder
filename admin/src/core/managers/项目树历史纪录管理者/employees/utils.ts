import {
  ProjectDiffDto,
  ProjectCreateDto,
  ProjectGroupCreateDto,
  ProjectUpdateWithIdDto,
  ProjectGroupUpdateWithIdDto,
} from '@/_gen/api';
import {
  DiffResult,
  ProjectStructureTreeDataNode,
  ProjectTreeNodeDataRecord,
} from '../../UIStoreManager';

/**
 * 将 DiffResult<ProjectStructureTreeDataNode> 转换为 ProjectDiffDto
 * @param diffResult - 比较树的结果
 * @param oldNodeDataRecord - key 到树节点数据的映射
 * @param newNodeDataRecord - key 到树节点数据的映射
 * @returns ProjectDiffDto
 */
export function convertDiffResultToProjectDiffDto(
  diffResult: DiffResult<ProjectStructureTreeDataNode>,
  oldNodeDataRecord?: ProjectTreeNodeDataRecord,
  newNodeDataRecord?: ProjectTreeNodeDataRecord,
): ProjectDiffDto {
  const projectsToCreate: ProjectCreateDto[] = [];
  const projectGroupsToCreate: ProjectGroupCreateDto[] = [];
  const projectsToUpdate: ProjectUpdateWithIdDto[] = [];
  const projectGroupsToUpdate: ProjectGroupUpdateWithIdDto[] = [];
  const projectIdsToDelete: number[] = [];
  const projectGroupIdsToDelete: number[] = [];

  diffResult.新增.forEach((新增操作) => {
    新增操作.recordItems.forEach((node) => {
      const recordItem = newNodeDataRecord![node.key];
      if (recordItem.type === 'file') {
        projectsToCreate.push({
          name: recordItem.title,
          projectGroupId: 新增操作.父节点key
            ? newNodeDataRecord![新增操作.父节点key].id
            : undefined,
          type: recordItem.projectFileType,
        });
      } else if (recordItem.type === 'folder') {
        projectGroupsToCreate.push({
          name: recordItem.title,
          parentGroupId: 新增操作.父节点key
            ? newNodeDataRecord![新增操作.父节点key].id
            : undefined,
        });
      }
    });
  });

  diffResult.更新?.forEach((更新操作) => {
    const newNode = 更新操作.newNode;
    const newNodeData = newNodeDataRecord![newNode.key];
    if (newNodeData.type === 'file') {
      projectsToUpdate.push({
        id: newNodeData.id,
        name: newNodeData.title,
        projectGroupId: 更新操作.节点key
          ? newNodeDataRecord![更新操作.节点key].id
          : undefined,
      });
    } else if (newNodeData.type === 'folder') {
      projectGroupsToUpdate.push({
        id: newNodeData.id,
        name: newNodeData.title,
        parentGroupId: 更新操作.节点key
          ? newNodeDataRecord![更新操作.节点key].id
          : undefined,
      });
    }
  });

  diffResult.删除.recordItems.forEach((node) => {
    const recordItem = oldNodeDataRecord![node.key];
    if (recordItem.type === 'file') {
      projectIdsToDelete.push(recordItem.id);
    } else if (recordItem.type === 'folder') {
      projectGroupIdsToDelete.push(recordItem.id);
    }
  });

  return {
    projectsToCreate,
    projectGroupsToCreate,
    projectsToUpdate,
    projectGroupsToUpdate,
    projectIdsToDelete,
    projectGroupIdsToDelete,
  };
}
