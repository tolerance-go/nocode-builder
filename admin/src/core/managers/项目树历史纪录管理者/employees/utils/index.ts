import {
  ProjectDiffDto,
  ProjectGroupUpdateWithIdDto,
  ProjectUpdateWithIdDto,
} from '@/_gen/api';
import {
  DiffResult,
  ProjectStructureTreeDataNode,
  ProjectTreeNodeDataRecord,
} from '../../../UIStoreManager';
import { handleAdditions } from './handleAdditions';

/**
 * 处理更新操作
 * @param diffResult - 比较树的结果
 * @param newNodeDataRecord - key 到树节点数据的映射
 * @returns 包含需要更新的项目和项目组的对象
 */
function handleUpdates(
  diffResult: DiffResult<ProjectStructureTreeDataNode>,
  newNodeDataRecord: ProjectTreeNodeDataRecord,
) {
  const projectsToUpdate: ProjectUpdateWithIdDto[] = [];
  const projectGroupsToUpdate: ProjectGroupUpdateWithIdDto[] = [];

  diffResult.更新?.forEach((更新操作) => {
    const newNode = 更新操作.newNode;
    const newNodeData = newNodeDataRecord[newNode.key];
    if (newNodeData.type === 'file') {
      projectsToUpdate.push({
        id: newNodeData.id,
        name: newNodeData.title,
        projectGroupId: 更新操作.节点key
          ? newNodeDataRecord[更新操作.节点key].id
          : undefined,
      });
    } else if (newNodeData.type === 'folder') {
      projectGroupsToUpdate.push({
        id: newNodeData.id,
        name: newNodeData.title,
        parentGroupId: 更新操作.节点key
          ? newNodeDataRecord[更新操作.节点key].id
          : undefined,
      });
    }
  });

  return { projectsToUpdate, projectGroupsToUpdate };
}

/**
 * 处理删除操作
 * @param diffResult - 比较树的结果
 * @param oldNodeDataRecord - key 到树节点数据的映射
 * @returns 包含需要删除的项目和项目组ID的对象
 */
function handleDeletions(
  diffResult: DiffResult<ProjectStructureTreeDataNode>,
  oldNodeDataRecord: ProjectTreeNodeDataRecord,
) {
  const projectIdsToDelete: number[] = [];
  const projectGroupIdsToDelete: number[] = [];

  diffResult.删除.recordItems.forEach((node) => {
    const recordItem = oldNodeDataRecord[node.key];
    if (recordItem.type === 'file') {
      projectIdsToDelete.push(recordItem.id);
    } else if (recordItem.type === 'folder') {
      projectGroupIdsToDelete.push(recordItem.id);
    }
  });

  return { projectIdsToDelete, projectGroupIdsToDelete };
}

/**
 * 将 DiffResult<ProjectStructureTreeDataNode> 转换为 ProjectDiffDto
 * @param diffResult - 比较树的结果
 * @param oldNodeDataRecord - key 到树节点数据的映射
 * @param newNodeDataRecord - key 到树节点数据的映射
 * @returns ProjectDiffDto
 */
export function convertDiffResultToProjectDiffDto(
  diffResult: DiffResult<ProjectStructureTreeDataNode>,
  oldNodeDataRecord?: ProjectTreeNodeDataRecord,
  newNodeDataRecord?: ProjectTreeNodeDataRecord,
): ProjectDiffDto {
  const additions = handleAdditions(diffResult, newNodeDataRecord!);
  const { projectsToUpdate, projectGroupsToUpdate } = handleUpdates(
    diffResult,
    newNodeDataRecord!,
  );
  const { projectIdsToDelete, projectGroupIdsToDelete } = handleDeletions(
    diffResult,
    oldNodeDataRecord!,
  );
  return {
    additions,
    projectsToUpdate,
    projectGroupsToUpdate,
    projectIdsToDelete,
    projectGroupIdsToDelete,
  };
}
