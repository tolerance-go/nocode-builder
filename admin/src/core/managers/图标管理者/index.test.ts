import { typedKeys } from '@/utils';
import { render } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import { 图标管理者 } from '.';

describe('图标管理者', () => {
  const manager = 图标管理者.getInstance();

  typedKeys(图标管理者.跟随组件id到组件映射).forEach((id) => {
    it(`根据id获取组件 - ${id}`, () => {
      const { container } = render(manager.根据id获取组件(id));
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
