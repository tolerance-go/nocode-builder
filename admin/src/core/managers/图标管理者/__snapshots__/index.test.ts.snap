// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`图标管理者 > 根据id获取组件 - 数据表项目节点 1`] = `
<span
  aria-label="table"
  class="anticon anticon-table"
  role="img"
>
  <svg
    aria-hidden="true"
    data-icon="table"
    fill="currentColor"
    focusable="false"
    height="1em"
    viewBox="64 64 896 896"
    width="1em"
  >
    <path
      d="M928 160H96c-17.7 0-32 14.3-32 32v640c0 17.7 14.3 32 32 32h832c17.7 0 32-14.3 32-32V192c0-17.7-14.3-32-32-32zm-40 208H676V232h212v136zm0 224H676V432h212v160zM412 432h200v160H412V432zm200-64H412V232h200v136zm-476 64h212v160H136V432zm0-200h212v136H136V232zm0 424h212v136H136V656zm276 0h200v136H412V656zm476 136H676V656h212v136z"
    />
  </svg>
</span>
`;

exports[`图标管理者 > 根据id获取组件 - 蓝图项目节点 1`] = `
<span
  aria-label="partition"
  class="anticon anticon-partition"
  role="img"
>
  <svg
    aria-hidden="true"
    data-icon="partition"
    fill="currentColor"
    focusable="false"
    height="1em"
    viewBox="64 64 896 896"
    width="1em"
  >
    <defs>
      <style />
    </defs>
    <path
      d="M640.6 429.8h257.1c7.9 0 14.3-6.4 14.3-14.3V158.3c0-7.9-6.4-14.3-14.3-14.3H640.6c-7.9 0-14.3 6.4-14.3 14.3v92.9H490.6c-3.9 0-7.1 3.2-7.1 7.1v221.5h-85.7v-96.5c0-7.9-6.4-14.3-14.3-14.3H126.3c-7.9 0-14.3 6.4-14.3 14.3v257.2c0 7.9 6.4 14.3 14.3 14.3h257.1c7.9 0 14.3-6.4 14.3-14.3V544h85.7v221.5c0 3.9 3.2 7.1 7.1 7.1h135.7v92.9c0 7.9 6.4 14.3 14.3 14.3h257.1c7.9 0 14.3-6.4 14.3-14.3v-257c0-7.9-6.4-14.3-14.3-14.3h-257c-7.9 0-14.3 6.4-14.3 14.3v100h-78.6v-393h78.6v100c0 7.9 6.4 14.3 14.3 14.3zm53.5-217.9h150V362h-150V211.9zM329.9 587h-150V437h150v150zm364.2 75.1h150v150.1h-150V662.1z"
    />
  </svg>
</span>
`;

exports[`图标管理者 > 根据id获取组件 - 视图项目节点 1`] = `
<span
  aria-label="build"
  class="anticon anticon-build"
  role="img"
>
  <svg
    aria-hidden="true"
    data-icon="build"
    fill="currentColor"
    focusable="false"
    height="1em"
    viewBox="64 64 896 896"
    width="1em"
  >
    <path
      d="M916 210H376c-17.7 0-32 14.3-32 32v236H108c-17.7 0-32 14.3-32 32v272c0 17.7 14.3 32 32 32h540c17.7 0 32-14.3 32-32V546h236c17.7 0 32-14.3 32-32V242c0-17.7-14.3-32-32-32zm-504 68h200v200H412V278zm-68 468H144V546h200v200zm268 0H412V546h200v200zm268-268H680V278h200v200z"
    />
  </svg>
</span>
`;

exports[`图标管理者 > 根据id获取组件 - 项目组文件夹 1`] = `
<span
  aria-label="folder"
  class="anticon anticon-folder"
  role="img"
>
  <svg
    aria-hidden="true"
    data-icon="folder"
    fill="currentColor"
    focusable="false"
    height="1em"
    viewBox="64 64 896 896"
    width="1em"
  >
    <path
      d="M880 298.4H521L403.7 186.2a8.15 8.15 0 00-5.5-2.2H144c-17.7 0-32 14.3-32 32v592c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V330.4c0-17.7-14.3-32-32-32zM840 768H184V256h188.5l119.6 114.4H840V768z"
    />
  </svg>
</span>
`;

exports[`图标管理者 > 根据id获取组件 - 项目组文件夹展开中 1`] = `
<span
  aria-label="folder-open"
  class="anticon anticon-folder-open"
  role="img"
>
  <svg
    aria-hidden="true"
    data-icon="folder-open"
    fill="currentColor"
    focusable="false"
    height="1em"
    viewBox="64 64 896 896"
    width="1em"
  >
    <path
      d="M928 444H820V330.4c0-17.7-14.3-32-32-32H473L355.7 186.2a8.15 8.15 0 00-5.5-2.2H96c-17.7 0-32 14.3-32 32v592c0 17.7 14.3 32 32 32h698c13 0 24.8-7.9 29.7-20l134-332c1.5-3.8 2.3-7.9 2.3-12 0-17.7-14.3-32-32-32zM136 256h188.5l119.6 114.4H748V444H238c-13 0-24.8 7.9-29.7 20L136 643.2V256zm635.3 512H159l103.3-256h612.4L771.3 768z"
    />
  </svg>
</span>
`;
